src/services/TestResultsService.ts

```diff
@@ -615,60 +444,47 @@ export class TestResultsService {
-              filteredTestTypeDates.push(moment(testType.testExpiryDate));
-            }
-          });
+    try {
+      const testResults = await this.getTestResultBySystemNumber({
+        systemNumber,
+        testStatus: enums.TEST_STATUS.SUBMITTED,
+        fromDateTime: new Date(1970, 1, 1),
+        toDateTime: new Date()
+      });
+      const filteredTestTypeDates: any[] = [];
+      testResults.forEach((testResult: { testTypes: any; vehicleType: any; vehicleSize: any; vehicleConfiguration: any; noOfAxles: any; }) => {
+        testResult.testTypes.forEach((testType: { testExpiryDate: string; testCode: string; }) => {
+          // prepare a list of annualTestTypes with expiry.
+          if (utils.ValidationUtil.isValidTestCodeForExpiryCalculation(testType.testCode.toUpperCase()) && DateProvider.isValidDate(testType.testExpiryDate)) {
+            filteredTestTypeDates.push(moment(testType.testExpiryDate));
+          }
         });
-        return filteredTestTypeDates;
-      }).then((annualTestTypeDates) => {
-        // fetch maxDate for annualTestTypes
-        if (annualTestTypeDates && annualTestTypeDates.length > 0) {
-          maxDate = moment.max(annualTestTypeDates).toDate();
-        }
-        return maxDate;
-      }).catch((err) => {
-        console.error("Something went wrong in generateExpiryDate > getMostRecentExpiryDateOnAllTestTypesBySystemNumber > getTestResults. Returning default test date and logging error:", err);
-        return maxDate;
       });
+      const annualTestTypeDates = filteredTestTypeDates;
+      // fetch maxDate for annualTestTypes
+      if (annualTestTypeDates && annualTestTypeDates.length > 0) {
+        maxDate = moment.max(annualTestTypeDates).toDate();
+      }
+      return maxDate;
+    } catch (err) {
+      console.error("Something went wrong in generateExpiryDate > getMostRecentExpiryDateOnAllTestTypesBySystemNumber > getTestResultBySystemNumber. Returning default test date and logging error:", err);
+      return maxDate;
+    }
   }
 
-  public getTestTypesWithTestCodesAndClassification(testTypes: Array<{ testTypeClassification: any; testTypeId: any; testCode?: any; }>, vehicleType: any, vehicleSize: any, vehicleConfiguration: any, noOfAxles: any,
-                                                    euVehicleCategory: any, vehicleClass: any, vehicleSubclass: any, numberOfWheelsDriven: any) {
+  public async getTestTypesWithTestCodesAndClassification(testTypes: Array<{ testTypeClassification: any; testTypeId: any; testCode?: any; }>, vehicleType: any, vehicleSize: any, vehicleConfiguration: any, noOfAxles: any,
+                                                          euVehicleCategory: any, vehicleClass: any, vehicleSubclass: any, numberOfWheelsDriven: any) {
     const promiseArray: any = [];
     if (testTypes === undefined) {
       testTypes = [];
```

Comment:
I think we could refactor how we create the testTypes

and remove some if and duplications
```
.then(currentTestCodesAndClassification => {
  testTypes[index].testTypeClassification = currentTestCodesAndClassification.testTypeClassification;
  testTypes[index].testCode = currentTestCodesAndClassification.defaultTestCode;
  if (currentTestCodesAndClassification.linkedTestCode) testTypes[index].testCode = currentTestCodesAndClassification.linkedTestCode;
})
```
We seem to be changing testTypeClassification and defaultTestCode if testTypes.length is 1 or more/less

Reply:
will do in 18782